### PR TITLE
UART DEM: Adjust to current OSD packet format

### PR DESCRIPTION
**Pull request by ThomasLeyk**
... and add cocotb tests for it.

### CODE REVIEW

File: modules/dem_uart/test/test_dem_uart.py [hunk 1]
```diff
@@ -156,3 +84,139 @@ def test_dem_uart_activation(dut):
...
+    read_thread  = cocotb.fork(dem_to_dii_rx(dut, num_transfers=1000))
+    
+    # read_thread implicitly waits for write thread to finish as well
+    yield read_thread.join()
+
+@cocotb.coroutine
+def dem_to_dii_tx(dut, num_transfers=500, max_delay=5, random_data=False):
+
+    for i in range(num_transfers):
+
```
Comment (koenenwmn) [inline modules/dem_uart/test/test_dem_uart.py]: Please remove some of the empty lines (like at the beginning of a loop) to keep the code a bit more compact.

File: modules/dem_uart/test/test_dem_uart.py [hunk 2]
```diff
@@ -156,3 +84,139 @@ def test_dem_uart_activation(dut):
...
+        for _ in range(delay):
+            yield RisingEdge(dut.clk)
+
+        data = (random.randint(0, 255) if random_data else 0x42)
+
+        _fifo.append(data)
+
+        dut.out_valid <= 0x1
+        dut.out_char <= int(data)
```
Comment (koenenwmn) [inline modules/dem_uart/test/test_dem_uart.py]: You don't need the 'int()' here since the variable already is an integer (also some other occasions).

File: modules/dem_uart/test/test_dem_uart.py [hunk 3]
```diff
@@ -156,3 +84,139 @@ def test_dem_uart_activation(dut):
...
+
+        # out_char & out_valid must be set until both out_ready & out_valid
+        # are asserted at the same rising edge, and for at least one clock cycle 
+        yield RisingEdge(dut.clk)
+
+        while not dut.out_ready.value or not dut.out_valid:
```
Comment (koenenwmn) [inline modules/dem_uart/test/test_dem_uart.py]: You don't need to check 'out_valid' here since you set it yourself. This way it implies that it might be cleared somewhere else.

File: modules/dem_uart/test/test_dem_uart.py [hunk 4]
```diff
@@ -156,3 +84,139 @@ def test_dem_uart_activation(dut):
...
+
+        for _ in range(delay):
+            yield RisingEdge(dut.clk)
+
+        dut.in_ready <= 0x1
+
+        # Again, make sure in_ready is asserted for at least one clock cycle
+        yield RisingEdge(dut.clk)
+
+        while not dut.in_valid.value or not dut.in_ready.value:
```
Comment (koenenwmn) [inline modules/dem_uart/test/test_dem_uart.py]: Same here you are setting 'in_ready' yourself, so you don't have to check for it.

File: modules/dem_uart/test/test_dem_uart.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+    dut._log.info("Check contents of MOD_CS")
+    yield access.assert_reg_value(MODULE_DI_ADDRESS, SENDER_DI_ADDRESS,
+                                  DiPacket.BASE_REG.MOD_CS.value, 16, 0)
+
+    yield _activate_module(dut)
+
+    dut._log.info("Check contents of MOD_CS")
+    yield access.assert_reg_value(MODULE_DI_ADDRESS, SENDER_DI_ADDRESS,
+                                  DiPacket.BASE_REG.MOD_CS.value, 16, 1)
```
Comment (imphil) [inline modules/dem_uart/test/test_dem_uart.py]: This test is incomplete: it only tests if the activation bit is stored, but not if the (de-)activation actually does anything. Does the module stop sending data if it has been deactivated? Does it start to send again when it has been activated? This behavior should be added to the spec, and then tested for.

If you don't want to do it now that's fine. In this case remove the test case, open an issue describing the problem, and add it to your backlog at https://github.com/orgs/opensocdebug/projects/1

File: modules/dem_uart/test/test_dem_uart.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+# Mirroring expected transfer through module
+_fifo = []
+
+@cocotb.test()
+def test_dem_to_dii(dut):
+    """
+    Verify that no chars are lost when the DII is busy
```
Comment (imphil) [inline modules/dem_uart/test/test_dem_uart.py]: This description is misleading: First, DII is an interface standard. The thing that's busy is the NoC, aka "Debug Interconnect (DI)". Second, you not only test if no chars are lost. You test the full transmission process.

File: modules/dem_uart/test/test_dem_uart.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+    yield _activate_module(dut)
+
+    write_thread = cocotb.fork(dem_to_dii_tx(dut, num_transfers=1000, random_data=True))
+    read_thread  = cocotb.fork(dem_to_dii_rx(dut, num_transfers=1000))
+    
+    # read_thread implicitly waits for write thread to finish as well
+    yield read_thread.join()
+
+@cocotb.coroutine
+def dem_to_dii_tx(dut, num_transfers=500, max_delay=5, random_data=False):
```
Comment (imphil) [inline modules/dem_uart/test/test_dem_uart.py]: Other coroutines (e.g. `_activate_module`) in this file have names starting with an underscore. Please follow this convention here as well.

File: modules/dem_uart/test/test_dem_uart.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+def test_dem_to_dii(dut):
+    """
+    Verify that no chars are lost when the DII is busy
+    """
+
+    yield _init_dut(dut)
+    yield _activate_module(dut)
+
+    write_thread = cocotb.fork(dem_to_dii_tx(dut, num_transfers=1000, random_data=True))
+    read_thread  = cocotb.fork(dem_to_dii_rx(dut, num_transfers=1000))
```
Comment (imphil) [inline modules/dem_uart/test/test_dem_uart.py]: Random testing is a very nice thing to explore corner cases, such as backpressure conditions. But to make sure that every test run covers those issues, you should add explicit tests for those cases to your test. E.g. add a test in which the NoC is never ready, and test what happens.  Even though this scenario is (most likely, depending on the random number generator) already covered, making corner cases explicit significantly helps during development.

File: modules/dem_uart/test/test_dem_uart.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    ex_packet = DiPacket();
+
+    for i in range(num_transfers):
+
+        delay = random.randint(1, max(1, max_delay))
+
+        for _ in range(delay):
+            yield RisingEdge(dut.clk)
+
+        rxPacket = yield reader.receive_packet(set_ready=True)
```
Comment (imphil) [inline modules/dem_uart/test/test_dem_uart.py]: The coding style in this file are identifiers with underscores, not camelCase ones: `rx_packet`

**Comment by ThomasLeyk** [comment]
With the last amend everything should be in order now.

**Comment by imphil** [comment]
Thanks @ThomasLeyk ! Could you combine the two commits into one (using git rebase -i), since they are not really separate any more. Then we can merge that and the next step is to continue with the specification PR.
